Ability to configure WebSocket connection

abodalevsky opened the pull request:
It's proposed to add options parameter to configure WebSocket. 
```javascript
const options = {
  headers: {
    origin: 'www.example.com'
  },
  proxy: {
    useDefault: true,
    uri: 'proxy.address.com',
    user: 'name',
    password: 'p@ssw0rd'
 }
}

const ws = new WebSocket('ws://echo.websocket.org', null, options)
```

**headers** section brings support _Origin_ header it addresses issue #253. Structure is aligned with implementation for Android.

**proxy** section extends support of proxy for UWP and brings support of proxy servers for WPF.
 - _useDefault_ - is relevant for WPF and indicates that connection should be established via proxy if it was auto-discovered. It is not relevant for UWP just because it's already behavior by default.
 - _uri_, _user_, _password_ - set credentials that use to authenticate to the proxy server through HTTP header authentication.

Any of this parameters can be omitted.

rozele commented:
@abodalevsky - can you link to the proxy implementation on Android? I don't see anything in https://github.com/facebook/react-native/blob/master/Libraries/WebSocket/WebSocket.js.

rozele commented:
Or in https://github.com/facebook/react-native/blob/master/ReactAndroid/src/main/java/com/facebook/react/modules/websocket/WebSocketModule.java for that matter.

abodalevsky commented:
I wasn't able to find it either. As I posted to discord, it uses OkHttpClient library capability, e.g. when proxy is detected WebSocket connection is tried to be established via proxy.

matthargett commented on ReactWindows/ReactNative.Shared.Tests/Modules/WebSocket/WebSocketOptionsTests.cs:
```diff
@@ -0,0 +1,150 @@
+}";
+            var jOptions = JObject.Parse(json);
+            var options = new WebSocketOptions(jOptions);
+            Assert.IsFalse(options.UseDefaultProxy);
```
Assert.That(options.UseDefaultProxy, Is.False);
// use the fluent style assertion syntax consistently throughout these new additions, please :)

abodalevsky commented:
@matthargett , thanks for your review. I've addressed feedback. In addition I have updated _flow type annotation_ for `options` parameter of WebSocket.

matthargett commented on ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs:
```diff
@@ -0,0 +1,62 @@
+            };
+
+            var proxy = WebRequest.DefaultWebProxy;
+            proxy.Credentials = CredentialCache.DefaultCredentials;
```
pass this value into the ctor as ICredentials:
https://msdn.microsoft.com/en-us/library/system.net.webproxy.credentials(v=vs.110).aspx

matthargett commented on ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs:
```diff
@@ -0,0 +1,62 @@
+                Scheme = address.Scheme == "wss" ? Uri.UriSchemeHttps : Uri.UriSchemeHttp
+            };
+
+            var proxy = WebRequest.DefaultWebProxy;
```
pass the webProxy into the ctor as IWebProxy so this class can be easily unit tested with mocks.

matthargett commented on ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs:
```diff
@@ -0,0 +1,62 @@
+                    return;
+                }
+
+                UserName = string.IsNullOrEmpty(defCreds.Domain) ? defCreds.UserName : $"{defCreds.Domain}\\{defCreds.UserName}";
```
use null coalescing operator.

abodalevsky replied:
there is check for `string.Empty` not for _null_

matthargett commented on ReactWindows/ReactNative.Net46/Modules/WebSocket/ProxyHelper.cs:
```diff
@@ -0,0 +1,62 @@
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException("options", "Parameter cannot be null");
```
use nameof(options) and null coalescing operator here

matthargett commented on ReactWindows/ReactNative.Net46/Modules/WebSocket/WebSocketModule.cs:
```diff
@@ -28,13 +28,7 @@ public WebSocketModule(ReactContext reactContext)
-                return "WebSocketModule";
-            }
-        }
+        public override string Name => "WebSocketModule";
```
👍
